File: Data Structures & Algorithms/count-number-of-islands/submission-2.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        seen=set()
        directions=[(0,1),(1,0),(-1,0),(0,-1)]
        r=len(grid)
        c=len(grid[0])
        cnt=0
        def valid(row,col):
            return 0<=row<r and 0<=col<c and grid[row][col]=='1'
        def dfs(row,col):
            for dx,dy in directions:
                new_row=row+dx
                new_col=col+dy
                if (new_row,new_col) not in seen and valid(new_row,new_col):
                    seen.add((new_row,new_col))
                    dfs(new_row,new_col)
        for i in range(r):
            for j in range(c):
                if (i,j) not in seen and grid[i][j]=='1':
                    seen.add((i,j))
                    cnt+=1
                    dfs(i,j)
        return cnt

File: Data Structures & Algorithms/count-paths/submission-2.java
class Solution {
    Map<String, Integer> memo;
    public boolean valid(int[][] grid,int row, int col){
        int r=grid.length;
        int c=grid[0].length;
        return ((0<=row)&&(row<r)&&(0<=col)&&(col<c));
    }
    public int dfs(int[][] grid,int m, int n, int row, int col){
        if ((row==m-1)&&(col==n-1)){
            return 1;
        }
        String key=row+","+col;
        if (memo.containsKey(key)){
            return memo.get(key);
        }
        int[][] directions=new int[][] {{0,1},{1,0}};
        for(int[] dir:directions){
            int newRow=row+dir[0];
            int newCol=col+dir[1];
            if (valid(grid,newRow,newCol)){
                memo.merge(key,dfs(grid,m,n,newRow,newCol),Integer::sum);
            }
        }
        return memo.get(key);
    }
    public int uniquePaths(int m, int n) {
        memo=new HashMap<>();
        int [][] grid= new int[m][n];
        return dfs(grid,m,n,0,0);
    }
}
